### PR TITLE
Lecture 1 setup and product dao

### PR DESCRIPTION
**Pull request by DashoGork**
- [x] Task 1.5: Implement ArrayListProductDao, Implement ArrayListProductDaoTest using junit4.
- [x] Task 1.6: Inject ArrayListProductDao into ProductListPageServlet, Implement ProductListServletTest.

### CODE REVIEW

File: src/main/java/com/es/phoneshop/exceptions/ProductNotFindException.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package com.es.phoneshop.exceptions;
+
+public class ProductNotFindException extends Throwable {
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/exceptions/ProductNotFindException.java]: Rename to ProductNotFoundException

Comment (nd524s) [reply src/main/java/com/es/phoneshop/exceptions/ProductNotFindException.java]: Don't extend  Throwable for custom exceptions.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 2]
```diff
@@ -1,25 +1,117 @@
...
+    public Product getProduct(Long id) throws ProductNotFindException {
+        readLock.lock();
+        try {
+            Product requiredProduct;
+            requiredProduct = products.stream().
+                    filter((product -> id.equals(product.getId()))).
+                    findFirst().orElseThrow(() -> new ProductNotFindException());
+            return requiredProduct;
+        } catch (ProductNotFindException e) {
+            throw new ProductNotFindException();
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: It's doesn't make sense to catch and re-throw the same exception.  Add meaningful message to the exception

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 3]
```diff
@@ -1,25 +1,117 @@
...
+        writeLock.lock();
+        try {
+            products.removeIf(product -> id.equals(product.getId()));
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void update(Product product) {
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Combine save and update methods in one

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 4]
```diff
@@ -1,25 +1,117 @@
...
+            readLock.unlock();
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+        readLock.lock();
+        try {
+            return products;
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: It should return products with non null price and stock level > 0

File: src/main/java/com/es/phoneshop/web/ProductListPageServlet.java [hunk 5]
```diff
@@ -1,41 +1,26 @@
...
-        result.add(new Product(9L, "nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
-        result.add(new Product(10L, "palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
-        result.add(new Product(11L, "simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        result.add(new Product(12L, "simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        result.add(new Product(13L, "simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/web/ProductListPageServlet.java:37]: Product pictures doesn't displayed on UI.

**Review by nd524s — CHANGES_REQUESTED** [review]

File: src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            if (id != null) {
+                Product requiredProduct;
+                requiredProduct = products.stream().
+                        filter((product -> id.equals(product.getId())
+                                && product.getPrice() != null
+                                && product.getStock() > 0)).
+                        findFirst().orElseThrow(() -> new ProductNotFoundException(id));
+                return requiredProduct;
+            } else
+                throw new InvalidParameterException();
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java]: Use IllegalArgumentException for this case

File: src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public List<Product> findProducts() {
+        readLock.lock();
+        try {
+            return products;
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java]: It should return products with non null price and stock level > 0

File: src/main/java/com/es/phoneshop/exceptions/ProductNotFoundException.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.es.phoneshop.exceptions;
+
+public class ProductNotFoundException extends Exception {
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/exceptions/ProductNotFoundException.java]: Extend RuntimeException

**Review by nd524s — CHANGES_REQUESTED** [review]

File: src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    public Product getProduct(Long id) throws ProductNotFoundException, InvalidParameterException {
+        readLock.lock();
+        try {
+            if (id != null) {
+                Product requiredProduct;
+                requiredProduct = products.stream().
+                        filter((product -> id.equals(product.getId())
+                                && product.getPrice() != null
+                                && product.getStock() > 0)).
+                        findFirst().orElseThrow(() -> new ProductNotFoundException(id));
```
Comment (nd524s) [inline src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java:40]: Throw exception with message. For instance "Product with id 1 doesn't exist "

**Review by nd524s — CHANGES_REQUESTED** [review]

File: src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java (hunk 9 above)
Comment (DashoGork) [reply src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java:40]: I have stasic string with exception message in exception class. Is it a bad practice and why?

Comment (nd524s) [reply src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java:40]: It's better to pass message directly in a place where it was thrown. It offeres better readability.

**Pull request merged**